library/src/main/java/com/datatorrent/lib/db/jdbc/AbstractJdbcPollInputOperator.java

```diff
@@ -438,119 +261,110 @@ public void endWindow()
+          windowId);
 
-      if (recoveredData != null) {
-        //skip the window and return if there was no incoming data in the window
-        if (recoveredData.left == null || recoveredData.right == null) {
-          return;
-        }
-
-        if (recoveredData.right.equals(rangeQueryPair.getValue()) || recoveredData.right.equals(previousUpperBound)) {
-          LOG.info("Matched so returning");
-          return;
-        }
+      if (recoveredData != null && shouldReplayWindow(recoveredData)) {
+        LOG.debug("[Recovering Window ID - {} for record range: {}, {}]", windowId, recoveredData.left,
+            recoveredData.right);
 
-        JdbcPollInputOperator jdbcPoller = new JdbcPollInputOperator();
-        jdbcPoller.setStore(store);
-        jdbcPoller.setKey(getKey());
-        jdbcPoller.setPartitionCount(getPartitionCount());
-        jdbcPoller.setPollInterval(getPollInterval());
-        jdbcPoller.setTableName(getTableName());
-        jdbcPoller.setBatchSize(getBatchSize());
-        isPollable = false;
-
-        LOG.debug("[Window ID -" + windowId + "," + recoveredData.left + "," + recoveredData.right + "]");
+        ps = store.getConnection().prepareStatement(
+            buildRangeQuery(rangeQueryPair.getKey(), (rangeQueryPair.getValue() - rangeQueryPair.getKey())),
+            TYPE_FORWARD_ONLY, CONCUR_READ_ONLY);
+        LOG.info("Query formed to recover data - {}", ps.toString());
 
-        jdbcPoller.setRangeQueryPair(new KeyValPair<String, String>(recoveredData.left, recoveredData.right));
+        emitReplayedTuples(ps);
 
-        jdbcPoller.ps = jdbcPoller.store.getConnection().prepareStatement(
-            JdbcMetaDataUtility.buildRangeQuery(jdbcPoller.getTableName(), jdbcPoller.getKey(),
-                jdbcPoller.getRangeQueryPair().getKey(), jdbcPoller.getRangeQueryPair().getValue()),
-            java.sql.ResultSet.TYPE_FORWARD_ONLY, java.sql.ResultSet.CONCUR_READ_ONLY);
-        LOG.info("Query formed for recovered data - {}", jdbcPoller.ps.toString());
+      }
 
-        emitReplayedTuples(jdbcPoller.ps);
+      if (currentWindowId == windowManager.getLargestRecoveryWindow()) {
+        try {
+          if (!isPollerPartition && rangeQueryPair.getValue() != null) {
+            ps = store.getConnection().prepareStatement(
+                buildRangeQuery(lastEmittedRecord, (rangeQueryPair.getValue() - lastEmittedRecord)), TYPE_FORWARD_ONLY,
+                CONCUR_READ_ONLY);
+          } else {
+            Integer bound = null;
+            if (lastEmittedRecord == null) {
+              bound = rangeQueryPair.getKey();
+            } else {
+              bound = lastEmittedRecord;
+            }
+            ps = store.getConnection().prepareStatement(buildRangeQuery(bound, Integer.MAX_VALUE), TYPE_FORWARD_ONLY,
+                CONCUR_READ_ONLY);
+          }
+          scanService.scheduleAtFixedRate(new DBPoller(), 0, pollInterval, TimeUnit.MILLISECONDS);
```

Comment:
If we don't start threads in "activate" method we want to do it in beginWindow once "reply" is done.